src/txdb.cpp

```diff
@@ -44,10 +47,21 @@ uint256 CCoinsViewDB::GetBestBlock() const {
     return hashBestChain;
 }
 
+uint256 CCoinsViewDB::GetUptoBlock() const {
+    uint256 hashBestChain;
+    if (!db.Read(DB_BEST_BLOCK_UPTO, hashBestChain))
+        return uint256();
+    return hashBestChain;
+}
+
 bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
     CDBBatch batch(db);
     size_t count = 0;
     size_t changed = 0;
+    size_t batch_size = (size_t)GetArg("-dbbatchsize", nDefaultDbBatchSize) << 20;
+    if (!hashBlock.IsNull()) {
+        batch.Write(DB_BEST_BLOCK_UPTO, hashBlock);
```

Comment:
That's a great point. I hadn't considered that in the case of a reorg the set of partially written changes may include things from multiple branches. It seems solvable by allowing the 'upto' blocks to be a list of tip hashes, and then at start time choose which ones to undo and which ones to replay. I think that's a problem for later, but it makes sense to have a comment about it.